File: Golang/usecase.go
package main

import "fmt"

// ABSTRACTION
type Employee interface {
    Role() string
    CalculateSalary() float64
    DisplayPayslip()
}

// ENCAPSULATION
type BaseEmployee struct {
    Name       string
    EmpID      string
    baseSalary float64   // private
}

func (b BaseEmployee) GetBaseSalary() float64 { return b.baseSalary }

// COMPOSITION
type FullTimeEmployee struct {
    BaseEmployee
    bonus float64
}

type PartTimeEmployee struct {
    BaseEmployee
    hoursWorked float64
}

type Contractor struct {
    BaseEmployee
    daysWorked float64
}

// FullTime methods
func (f FullTimeEmployee) Role() string { return "Full Time Employee" }
func (f FullTimeEmployee) CalculateSalary() float64 {
    return f.GetBaseSalary() + f.bonus
}
func (f FullTimeEmployee) DisplayPayslip() {
    fmt.Printf("\n %s (%s) | %s | ₹%.0f\n",
        f.Name, f.EmpID, f.Role(), f.CalculateSalary())
}

// PartTime methods
func (p PartTimeEmployee) Role() string { return "Part Time Employee" }
func (p PartTimeEmployee) CalculateSalary() float64 {
    return p.GetBaseSalary() * p.hoursWorked
}
func (p PartTimeEmployee) DisplayPayslip() {
    fmt.Printf("\n💼 %s (%s) | %s | ₹%.0f\n",
        p.Name, p.EmpID, p.Role(), p.CalculateSalary())
}

// Contractor methods
func (c Contractor) Role() string { return "Contractor" }
func (c Contractor) CalculateSalary() float64 {
    return c.GetBaseSalary() * c.daysWorked
}
func (c Contractor) DisplayPayslip() {
    fmt.Printf("\n💼 %s (%s) | %s | ₹%.0f\n",
        c.Name, c.EmpID, c.Role(), c.CalculateSalary())
}

func usecase() {
    // POLYMORPHISM
    employees := []Employee{
        FullTimeEmployee{BaseEmployee: BaseEmployee{Name: "Ravi", EmpID: "E001", baseSalary: 50000}, bonus: 10000},
        PartTimeEmployee{BaseEmployee: BaseEmployee{Name: "Sara", EmpID: "E002", baseSalary: 500}, hoursWorked: 80},
        Contractor{BaseEmployee: BaseEmployee{Name: "John", EmpID: "E003", baseSalary: 3000}, daysWorked: 20},
    }

    for _, e := range employees {
        e.DisplayPayslip()
    }
}

File: Python/usecase.py
from abc import ABC, abstractmethod

class Employee(ABC):
    def __init__(self, name, emp_id, base_salary):
        self.name = name
        self.emp_id = emp_id
        self.__base_salary = base_salary    # Encapsulation

    def get_base_salary(self):
        return self.__base_salary

    @abstractmethod
    def calculate_salary(self):
        pass

    @abstractmethod
    def role(self):
        pass

    def display_payslip(self):
        print(f"""
💼 Payslip:
   ID           : {self.emp_id}
   Name         : {self.name}
   Role         : {self.role()}
   Base Salary  : ₹{self.__base_salary}
   Total Salary : ₹{self.calculate_salary()}
        """)


# INHERITANCE
class FullTimeEmployee(Employee):
    def __init__(self, name, emp_id, base_salary, bonus):
        super().__init__(name, emp_id, base_salary)
        self.__bonus = bonus

    def role(self):
        return "Full Time Employee"

    def calculate_salary(self):
        return self.get_base_salary() + self.__bonus


class PartTimeEmployee(Employee):
    def __init__(self, name, emp_id, hourly_rate, hours_worked):
        super().__init__(name, emp_id, hourly_rate)
        self.__hours_worked = hours_worked

    def role(self):
        return "Part Time Employee"

    def calculate_salary(self):
        return self.get_base_salary() * self.__hours_worked


class Contractor(Employee):
    def __init__(self, name, emp_id, daily_rate, days_worked):
        super().__init__(name, emp_id, daily_rate)
        self.__days_worked = days_worked

    def role(self):
        return "Contractor"

    def calculate_salary(self):
        return self.get_base_salary() * self.__days_worked


# POLYMORPHISM
employees = [
    FullTimeEmployee("Ravi", "E001", 50000, 10000),
    PartTimeEmployee("Sara", "E002", 500, 80),
    Contractor("John", "E003", 3000, 20),
]

for emp in employees:
    emp.display_payslip()    # Same method — different salary calculation